Framework/DataHandling/src/SetSample.cpp

```diff
@@ -405,6 +588,109 @@ const Geometry::SampleEnvironment *SetSample::setSampleEnvironment(
   return &(exptInfo.sample().getEnvironment());
 }
 
+/**
+ * Set the requested sample environment from shape XML string
+ * @param exptInfo A reference to the ExperimentInfo to receive the environment
+ * @param canGeomArgs The dictionary of flags for the environment
+ * @param canMaterialArgs The dictionary of material parameters
+ * @return A pointer to the new sample environment
+ */
+const Geometry::SampleEnvironment *SetSample::setSampleEnvironmentFromXML(
+    API::ExperimentInfo &exptInfo,
+    const Kernel::PropertyManager_const_sptr &canGeomArgs,
+    const Kernel::PropertyManager_const_sptr &canMaterialArgs) {
+  const auto refFrame = exptInfo.getInstrument()->getReferenceFrame();
+  const auto xml = tryCreateXMLFromArgsOnly(*canGeomArgs, *refFrame);
+  if (!xml.empty()) {
+    ShapeFactory sFactory;
+    // Create the object
+    auto shape = sFactory.createShape(xml);
+    if (shape->hasValidShape()) {
+      if (canMaterialArgs) {
+        ReadMaterial::MaterialParameters materialParams;
+        setMaterial(materialParams, *canMaterialArgs);
+        if (materialParams.sampleVolume <= 0.) {
+          materialParams.sampleVolume = shape->volume() * CUBIC_METRE_TO_CM;
+        }
+        ReadMaterial reader;
+        reader.setMaterialParameters(materialParams);
+        auto canMaterial = reader.buildMaterial();
+        shape->setMaterial(*canMaterial);
```

Comment:
That's a possibility; I did not change that again for not accidentally breaking the API, since the validation that I am doing here is far more strict. Actually it would be neater to use the child algorithm for both, but I'm not sure if SetSampleMaterial can handle the container.